### ISSUE TITLE
OpaqueKeyDecrypter for JWE

### ISSUE THREAD

**Issue by ilya-korotya** [lestrrat-go/jwx#924]
**Abstract**
Hi. I want to use a hardware repository with a private key for decrypting JWE messages. I don't want the key from the hardware repository to escape from the hardware repository. I tried using the [KeyProvider](https://github.com/lestrrat-go/jwx/blob/develop/v2/examples/jwe_example_test.go#L84-L108), but it requires me to save the private key to the sink variable. The JOSE library provides the [OpaqueKeyDecrypter](https://github.com/go-jose/go-jose/blob/v2.6.1/opaque.go#L125-L127) for custom implementation of decryption logic.
Is it possible to do something similar in the JWE library?

**Describe the proposed solution/change**
My vision is that we can pass our own decryption interface to the `jwe.WithKey()` function.
```go
type HardWare {
 s Storage
}

func (hw *HW) Decrypt(decryptedKey []byte, opts interface{}) ([]byte, error) {
  kid := opts.GetKid()
  return s.Decrypt(decryptedKey, kid)
}

func main() {
   s := HardWare{}
   privkey, encrypted, _ := exampleGenPayload()
   decrypted, _ := jwe.Decrypt(encrypted, jwe.WithKey(jwa.RSA1_5, s))
}
```

**Comment by lestrrat** [comment]
@ilya-korotya this library supports using `crypto.Signer`, does that work for you? (I just checked, and I realize the documentation is lacking... it's only mentioned in [02-jws.md](https://github.com/lestrrat-go/jwx/blob/develop/v2/docs/02-jws.md#using-cloud-kms-services))

Documentation is a bit outdated, but you should be able to do [what this PoC does](https://github.com/jwx-go/crypto-signer).

Let me know if that doesn't work for you.

**Comment by lestrrat** [comment]
Sorry, ignore my previous post (which I deleted)

**Comment by lestrrat** [comment]
I see, so you want a pluggable decrypter. That will involve some internal refactoring. Let me think about it.

**Comment by lestrrat** [comment]
mental note: you would need a Decrypter _and_ an Encrypter.

**Comment by lestrrat** [comment]
Okay, @ilya-korotya want to try #925 ?

**Comment by ilya-korotya** [comment]
Thank you so much for the quick response! I tried out this PR, and it's very good! Here's my unit test using a mocked KMS provider.

```go
package anoncrypt

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"fmt"
	"testing"

	"github.com/lestrrat-go/jwx/v2/jwa"
	"github.com/lestrrat-go/jwx/v2/jwe"
	"github.com/stretchr/testify/require"
)

type KMS struct {
	pk *rsa.PrivateKey
}

func (kms *KMS) DecryptKey(
	alg jwa.KeyEncryptionAlgorithm,
	encryptedKey []byte,
	recipient jwe.Recipient,
	message *jwe.Message,
) ([]byte, error) {
	return rsa.DecryptOAEP(sha256.New(), rand.Reader, kms.pk, encryptedKey, nil)
}

func TestAnoncrypt(t *testing.T) {
	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatal(err)
	}

	payload := []byte("Lorem Ipsum")
	encrypted, err := jwe.Encrypt(
		payload,
		jwe.WithCompact(),
		jwe.WithKey(jwa.RSA_OAEP_256, privKey.PublicKey),
		jwe.WithContentEncryption(jwa.A128GCM),
	)
	require.NoError(t, err, `jwe.Encrypt should succeed`)
	fmt.Println("token:", string(encrypted))

	raw, err := jwe.Decrypt(encrypted, jwe.WithKey(jwa.RSA_OAEP_256, &KMS{privKey}))
	require.NoError(t, err)
	fmt.Println("raw:", string(raw))
}

```

**Comment by lestrrat** [comment]
@ilya-korotya Cool. Can you play around with it and think for a few days if this really fills your needs and get back to me? I'm more less okay with the design, but as I don't use this feature myself I'm not too sure if I missed anything, so would appreciate feedback from somebody actually using it.

I'm particularly interested in knowing if the API for `EncryptKey()` and `DecryptKey()` are sufficient (or maybe too much) in terms of the parameters they are expecting.

**Comment by ilya-korotya** [comment]
@lestrrat sure, I will test this PR and let you know the results. Thank you!

**Comment by lestrrat** [comment]
@ilya-korotya JFYI, I'll probably make a release _without_ this feature soon, unless I get some input about the API. This feature probably won't die, but it will definitely be punted to some other time

**Comment by ilya-korotya** [comment]
Hi @lestrrat. This API looks suitable for me. I implemented what I wanted. But I'm working on testing these changes... Maybe in the future, I will have some proposals about API. But currently it looks pretty cool.

**Comment by lestrrat** [comment]
fixed by #925
